### PR TITLE
Obfuscation

### PR DESCRIPTION
**Pull request by jnfeinstein**

### CODE REVIEW

**Comment by jnfeinstein** [comment]
I am assuming it is desirable for RuboCop to pass successfully?

**Comment by johnnyshields** [comment]
@jnfeinstein for obfuscate, how does it actually work? Maybe instead of `obfuscate` we should allow `format`-ing values (e.g. with a proc) so its a more general use case?

**Comment by jnfeinstein** [comment]
Current iteration replaces the value with '********'.  `format` options sounds fine.  Config would have to be a hash of fields instead of an array, normalized by db name internally.  Then call the proc with the value if proc, or inject the value into the string if string (i.e. `'123 %s 456' % val` which works fine when the string has no placeholders).

The config would be a bit uglier for obfuscation since you'd have to add the obfuscation string for each field rather than using an array.

**Comment by dblock** [comment]
This is a great start, but I definitely think we want to think harder about the DSL here. Do we like Rails `filter_parameters`?

I vote for a generic transform. I would consider going further and model this as events on write and read. You could have global callbacks for any model/field (implement obfuscation on anything called "password" for example) or have local callbacks for individual fields.

Want to give it a try @jnfeinstein?

**Comment by johnnyshields** [comment]
@dblock `format` or `filter` is ok with me.

File: .gitignore [hunk 1]
```diff
@@ -44,3 +44,6 @@ pkg
 #
 # For vim:
 #*.swp
+
+# JetBrains
+.idea
```
Comment (johnnyshields) [inline .gitignore]: You can set this in your local global gitignore

**Comment by johnnyshields** [comment]
Considering the DSL, it would make sense to have both `format` and `hide` options. Hide would be simpler.

```
track_history on: :recipes, hide: { recipes: [:secret_ingredient] }, format: { recipes: { price: ->(val){ "$#{val}" } } }
```

File: .gitignore (hunk 1 above)
Comment (jnfeinstein) [inline .gitignore]: TIL 👍

**Comment by jnfeinstein** [comment]
@dblock `filter_parameters` is exactly what I'm trying to accomplish.  I tried to match the syntax of the `on` option.  Could you elaborate on this event system?  The `format` option that @johnnyshields has suggested seems fairly flexible without unnecessary complication.

@johnnyshields I fear `hide` might be too close to `except` (both connote getting rid of something).  `format` will do just fine.

**Comment by jnfeinstein** [comment]
Updated w/ `rubocop` fixes, removed gitignore changes.

**Comment by jnfeinstein** [comment]
@dblock @johnnyshields changed `obfuscate` to `format`.  Configuration requires hashes (all the way down) as in `{ foo: '%s', bar: { baz: ->(v){ v.to_s * 2 }`.

If this is solid then I'll update README and CHANGELOG to match.

**Comment by dblock** [comment]
I like it.

**Comment by johnnyshields** [comment]
I like it too. One question: `foo: '%s'` does this automatically use string interpolation or force the string literal `"%s"`?

**Comment by jnfeinstein** [comment]
@johnnyshields automatically uses interpolation if the value is a string.

**Comment by jnfeinstein** [comment]
`rubocop` is throwing the following errors:

```
lib/mongoid/history/options.rb:3:5: C: Class has too many lines. [120/103]
    class Options
    ^^^^^
spec/unit/attributes/base_spec.rb:17:5: C: Missing top-level class documentation comment.
    class ModelTwo
    ^^^^^
```

Neither seems actionable, agreed?

**Comment by dblock** [comment]
For Rubocop run `rubocop -a` then `rubocop --auto-gen-config` for exceptions.

**Comment by jnfeinstein** [comment]
Great.  Added the failing styles as exclusions.  The options style error is presumably valid but I don't believe it's within the scope of this PR to refactor to make rubocop happy.

**Comment by dblock** [comment]
@johnnyshields This is all yours to merge!

**Comment by johnnyshields** [comment]
Merged. Really nice work @jnfeinstein

**Pull request merged**